mods/d2k/maps/harkonnen-08/harkonnen08.lua

```diff
@@ -0,0 +1,328 @@
+}
+
+OrdosAttackDelay =
+{
+	easy = DateTime.Minutes(3),
+	normal = DateTime.Minutes(2) + DateTime.Seconds(20),
+	hard = DateTime.Minutes(1)
+}
+
+MercenaryStarportDelay = DateTime.Minutes(1) + DateTime.Seconds(20)
+
+OrdosAttackWaves =
+{
+	easy = 4,
+	normal = 5,
+	hard = 6
+}
+
+InitialOrdosReinforcements =
+{ 
+	{ "trooper", "trooper", "trooper", "trooper", "light_inf", "light_inf", "light_inf" },
+	{ "trooper", "trooper", "trooper", "trooper", "trooper", "combat_tank_o", "combat_tank_o" }
+}
+
+InitialAtreidesReinforcements = { "combat_tank_a", "combat_tank_a", "quad", "quad", "trike" }
+
+InitialMercenaryReinforcements = { "trooper", "trooper", "trooper", "trooper", "quad", "quad" }
+
+OrdosPaths =
+{
+	{ OrdosEntry1.Location, OrdosRally1.Location },
+	{ OrdosEntry2.Location, OrdosRally2.Location },
+	{ OrdosEntry3.Location, OrdosRally3.Location },
+	{ OrdosEntry4.Location, OrdosRally4.Location }
+}
+
+InitialOrdosPaths =
+{
+	{ OLightFactory.Location, OrdosRally5.Location },
+	{ OHiTechFactory.Location, OrdosRally6.Location }
+}
+
+SaboteurPaths =
+{
+	{ SaboteurWaypoint1.Location, SaboteurWaypoint2.Location, SaboteurWaypoint3.Location },
+	{ SaboteurWaypoint4.Location, SaboteurWaypoint5.Location, SaboteurWaypoint6.Location }
+}
+
+InitialAtreidesPath = { AStarport.Location, AtreidesRally.Location }
+
+InitialMercenaryPath = { MStarport.Location, MercenaryRally.Location }
+
+SendStarportReinforcements = function(faction)
+	Trigger.AfterDelay(MercenaryStarportDelay, function()
+		if MStarport.IsDead or MStarport.Owner ~= faction then
+			return
+		end
+		
+		reinforcements = Utils.Random(MercenaryStarportReinforcements)
+
+		local units = Reinforcements.ReinforceWithTransport(faction, "frigate", reinforcements, { MercenaryStarportEntry.Location, MStarport.Location + CVec.New(1, 1) }, { MercenaryStarportExit.Location })[2]
+		Utils.Do(units, function(unit)
+			unit.AttackMove(MercenaryAttackLocation)
+			IdleHunt(unit)
+		end)
+
+		SendStarportReinforcements(faction)
+	end)
+end
+
+SendAirStrike = function()
+	if AHiTechFactory.IsDead or AHiTechFactory.Owner ~= atreides_enemy then
+		return
+	end
+
+	local targets = Utils.Where(player.GetActors(), function(actor)
+		return
+			actor.HasProperty("Sell") and
+			actor.Type ~= "wall" and
+			actor.Type ~= "medium_gun_turret" and
+			actor.Type ~= "large_gun_turret" and
+			actor.Type ~= "wind_trap"
+	end)
+
+	if #targets > 0 then
+		AHiTechFactory.SendAirstrike(Utils.Random(targets).CenterPosition, true, 0)
+	end
+
+	Trigger.AfterDelay(DateTime.Minutes(5), SendAirStrike)
+end
+
+BuildSaboteur = function()
+	if OPalace.IsDead or OPalace.Owner ~= ordos then
+		return
+	end
+
+	local targets = Utils.Where(player.GetActors(), function(actor)
+		return
+			actor.HasProperty("Sell") and
+			actor.Type ~= "wall"
+	end)
+
+	if #targets > 0 then
+		OPalace.Produce("saboteur")
+
+		Trigger.OnProduction(OPalace, function(producer, produced)
```

Comment:
This is creating a new trigger without removing the old each time this function runs! (Which meant that you still called `SendSaboteur` with the old `targets`...)

I'd suggest rewriting the entire section to avoid this:
```

GetSaboteurTargets = function(player)
	return Utils.Where(player.GetActors(), function(actor)
		return actor.HasProperty("Sell") and actor.Type ~= "wall"
	end)
end

BuildSaboteur = function()
	if OPalace.IsDead or OPalace.Owner ~= ordos then
		return
	end

	local targets = GetSaboteurTargets(player)
	if #targets > 0 then
		local saboteur = Actor.Create("saboteur", true, { Owner = ordos, Location = OPalace.Location + CVec.New(0, 2) })
		saboteur.Move(saboteur.Location + CVec.New(0, 1))
		saboteur.Wait(DateTime.Seconds(5))

		local path = Utils.Random(SaboteurPaths)
		saboteur.Move(path[1])
		saboteur.Move(path[2])
		saboteur.Move(path[3])

		SendSaboteur(saboteur)
	end

	Trigger.AfterDelay(DateTime.Minutes(1) + DateTime.Seconds(30), BuildSaboteur)
end

SendSaboteur = function(saboteur)
	local targets = GetSaboteurTargets(player)
	if #targets < 1 then
		return
	end

	local target = Utils.Random(targets)
	saboteur.Demolish(target)

	-- 'target' got removed from the world in the meantime
	saboteur.CallFunc(function()
		SendSaboteur(saboteur)
	end)
end
```

This has the bonus of saboteurs acquiring a new target when the old one was removed from the world. (Except when the player had no base buildings at all at one moment. I thought it's not worth covering that case however. What do you think?)